### DISCUSSION TITLE
Preview feature feedback: `@prisma/adapter-d1` / Cloudflare D1 support

### DISCUSSION THREAD

**Discussion by janpio** [prisma/prisma#23646] [Feedback]
Support for the Cloudflare D1 is in preview behind the `driverAdapters` preview feature with the release of 5.12.0 on April 2nd.

Please share your feedback about `@prisma/adapter-d1` in this discussion.

- If you encounter a bug, please [open a bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=bug_report.md&title=) in this repo.
- If the feature is working well for you, please share this in a comment below or leave a 👍 on this issue.
- If you have any questions, don't hesitate to ask them in the [Prisma Discord](https://pris.ly/discord) community

**Comment by m-dierker** [comment]
Feedback: The Prisma engine adds about 2MB to the Cloudflare D1 bundle size. This means that free tier users (with a 1MB bundle size limit) are unable to use Prisma without confusing errors. 😜 The minimum cost to use Prisma becomes $60/yr. I wound up switching to Drizzle for my project.

Most of the engine seems to be a 2MB seems to be a giant webassembly file. Cloudflare doesn't count static assets against the 1MB limit. I wonder if there's a way to load it in asynchronously as a static asset rather than compiling it into the function.

**Reply by janpio** [reply]
Yes, this is an unfortunate limitation on our side right now. The Webassembly file actually clocks in at ~785kb gzipped, which is what Cloudflare counts towards the bundle size limit. So small apps can actually be deployed in the free account.

But good you found a solution that works for you. As soon as we make progress here, we will let you know and you can consider switching back if that is still an option then. Thanks!

(The Webassembly file unfortunately has to be part of the function as far as I know, as it is loaded by the function and _executed_ - which is not allowed for static assets. But thanks for mentioning this idea anyway.)

**Reply by janpio** [reply]
A related issue: https://github.com/prisma/prisma/issues/23350

**Comment by hrueger** [comment]
Feedback: It's working generally well for me, I have a couple of smaller applications deployed already using the D1 adapter. However, my main application has more than 100 rows in almost every table and I'm running into trouble here:
I'm usually not someone to ping people or constantly beg for issues to be resolved, but I'd really appreciate if someone took a look at #23743. This is quite a bummer since it affects almost every query in that app (you only need more than 100 rows and select with one relation to trigger that bug). I believe this affects lots of people once they start running production apps.
Thanks in advance!

**Reply by hrueger** [reply]
Update: I've found a hacky workaround for that bug by inlining escaped variables if there are more than 100. However, now that I'm actually running the app with data on Cloudflare Workers, I noticed some performance issues. On my VPS, page load times were about 400-600ms in total, now we're looking at 1800 - 2500ms... I'm trying to figure out where this is coming from, but I suspect prisma to be the culprit. I tried to implement Tracing with OpenTelemetry to see query duration, but that does not work on Workers due to it using node apis like `process`, etc

**Reply by hrueger** [reply]
I built a super simple test app which deletes a row and recreates it 100 times. That means, I do 200 round trips to the database. Here are the request durations:
- Locally with node and miniflare: 500 - 600ms
- Locally with wrangler dev and miniflare: 150 - 160ms
- On Workers: 7500 - 7800ms
That said, now it looks like the D1 database itself might be so slow? The metrics in the dashboard show a constant latency of 0ms.

Edit: It indeed looks like this. I patched the d1 adapter like so:
```js
      let start = performance.now();
      const stmt = this.client.prepare(query.sql).bind(...query.args);
      console.log(`[js::performIO] prepare/bind took ${performance.now() - start}ms`);
      start = performance.now();
      if (executeRaw) {
        const r = (0, import_driver_adapter_utils2.ok)(await stmt.run());
        console.log(`[js::performIO] run took ${performance.now() - start}ms`);
        return r;
      } else {
        const [columnNames, ...rows] = await stmt.raw({ columnNames: true });
        const r = (0, import_driver_adapter_utils2.ok)([columnNames, rows]);
        console.log(`[js::performIO] raw took ${performance.now() - start}ms`);
        return r;
      }
```
It prints:
```
  (log) [js::performIO] prepare/bind took 0ms
  (log) [js::performIO] raw took 30ms
  (log) [js::performIO] ok took 0ms
```
(note that prepare/bind and ok will always be 0, because [timers only advance after io](https://developers.cloudflare.com/workers/runtime-apis/performance/#performancenow)).
~~According to [this chart](https://blog.cloudflare.com/d1-turning-it-up-to-11#turning-it-up-to-11) it should be more like 1.8ms...~~ (that article is about sql processing time, not network latency / round trip time).

Edit 2:
The problem has been solved. "Smart Placing" improves the latency to D1 massively and fixes the issue.

**Reply by steebchen** [reply]
Thanks for the "smart placing" tip!

**Comment by derekhearst** [comment]
I often write queries with many to many relationships being included, and it errors most of the time due to D1 having a very small parameter max length.

I would really like to see joins being supported so I don't have to query raw.

I am really happy with the performance.

**Reply by janpio** [reply]
Are there related issues tracking this problem? If so, please link to them here. If not, please open one. Thanks!

Probably related, but not an issue for the actual problem this is causing: https://github.com/prisma/prisma/issues/23348#issuecomment-2071797735

**Comment by jamiehaywood** [comment]
feedback around the D1 developer workflow:

Generally really nice & well thoughout. I have a use-case where my local `.wrangler/../miniflare-D1DatabaseObject/xx.sqlite` is in the root of my monorepo, so I'm not able to run the suggested prisma migrate script - `npx prisma migrate diff --from-local-d1 --to-schema-datamodel ./schema.prisma --script --output migrations/0002_second_migration.sql`. I think the `--from-local-d1` should accept an optional argument to point to the `.wrangler` folder.

**Reply by janpio** [reply]
Oh monorepos, of course. Thanks for mentioning that! 

Can you maybe quickly create a feature request issue with this? Then we will put that into the backlog for our D1 support.

**Reply by jackshunn** [reply]
Did a feature request get made for this? I've been using an alternative setup for a monorepo that might be often be simpler than adding more command line arguments to `prisma migrate`. I'd be happy to contribute my idea to the conversation.

In short, the idea is to have a distinct node.js package somewhere in the monorepo with only `wrangler` and `primsa` as dev dependencies. It's responsibilities are to manage D1 migrations and export the generated prisma client.

**Comment by alex-mironov** [comment]
Do I understand correctly that in each loader I'd need to create a new prisma client, like this?
```typescript
  const adapter = new PrismaD1(env.DB)
  const prisma = new PrismaClient({ adapter })
```

Of course I can add a function to make it reusable, but is it how it's supposed to be used? What about `$connect/$disconnect`, should I care about them?

```typescript
export const getPrisma = (env: Env) => {
  const adapter = new PrismaD1(env.DB)
  const prisma = new PrismaClient({ adapter })
  return prisma
}
```

Are there any disadvantages of such approach? Is there a way to create a single instance of `PrismaClient` and use it across the app?

**Reply by stewarthsoj** [reply]
I ended up initializing the Prisma client in some middleware and sharing the client via the request context.

**Reply by Noitidart** [reply]
Why would you need to share via request context @JoshStwrt ? A simple util to import `getPrisma` seems cleaner solution no?

**Comment by zshannon** [comment]
Something's exploding here on inserts and it's killing performance. Here are some benchmarks from inserting 10, 100, 1000 new rows using `prisma.createMany`, `prisma.$executeRawUnsafe`, and `d1.exec`. `prisma.$executeRawUnsafe` and `d1.exec` are close enough, but `prisma.createMany` blows up with row count.

`Running on Worker` (with region hint `wnam`, making requests from SF, CA, USA)
(median of 10 runs) | Prisma.createMany ms | Prisma.$executeRawUnsafe ms | D1.exec ms
-- | -- | -- | --
10 records | 33 | 33 | 32
100 recrods | 98 | 35.5 | 35
1000 records | 778.5 | 42.5 | 46

`Running on localhost` (M1 Max)
(median of 10 runs) | Prisma.createMany ms | Prisma.$executeRawUnsafe ms | D1.exec ms
-- | -- | -- | --
10 records | 9 | 5 | 2
100 recrods | 19 | 4 | 2
1000 records | 81 | 9 | 4

You can test the worker version [here](https://performance-test.prisma-d1-perf-benchmark.pages.dev/), which is running the code from this [repo/branch](https://github.com/zshannon/prisma-d1-perf-benchmark/blob/f8a93ba22dd6f8c19f26f8ba7c0a46b680741385/src/app/actions.ts)
```typescript
export async function writeProperties(count: number) {
	const t0 = performance.now()
	const data = Array.from({ length: count })
		.fill(null)
		.map(() => ({ key: createId(), value: createId() }))
	const t1 = performance.now()
	await prisma.property.createMany({
		data,
	})
	const t2 = performance.now()
	const values = Array.from({ length: count })
		.fill(null)
		.map(() => `('${createId()}', '${createId()}')`)
		.join(', ')
	const sql = `INSERT INTO Property (key, value) VALUES ${values}`
	const t3 = performance.now()
	await prisma.$executeRawUnsafe(sql)
	const t4 = performance.now()
	const values2 = Array.from({ length: count })
		.fill(null)
		.map(() => `('${createId()}', '${createId()}')`)
		.join(', ')
	const sql2 = `INSERT INTO Property (key, value) VALUES ${values2}`
	const t5 = performance.now()
	await process.env.DB.exec(sql2)
	const t6 = performance.now()
	return {
		'D1.exec': t6 - t5,
		'Prisma.createMany': t2 - t1,
		'Prisma.$executeRawUnsafe': t4 - t3,
	}
}
```

```
"@prisma/adapter-d1": "5.21.0",
"@prisma/client": "5.21.0",
```

**Reply by zshannon** [reply]
I should add: this benchmark is for 2 columns of 24 character strings. I originally ran into the performance issue for an application that stores Google Calendar events in rows that are 40 columns (~2000 characters of text) wide, and in that case `upsert` on just 100 rows takes __30 seconds__!

**Comment by nicolaspapp** [comment]
I have feedback regarding the D1 adapter. From what I am seeing in the logs, nested reads are not being done on either the `include` or `select` statements. I have even tried adding `relationJoins` as preview feature but when I see the logs it is still doing the extra queries. 
I believe this could be very important for having performant queries. Smaller priority but nested writes would also be handy

**Comment by codercatdev** [comment]
When developing locally using `--from-local-d1` it would be great to allow a source reference and not auto find this file. The reasoning behind this is that I centralize many workers to a single D1 database using `--persist-to` so the database is not in the same location that I am running something like `pnpx prisma migrate diff --from-local-d1 --to-schema-datamodel ./prisma/schema.prisma --script --output migrations/0002_create_user_table.sql`.

Not sure if I should open an issue?

**Comment by disturbing** [comment]
General feedback regarding the adapter:

1. Although transactions aren't supported, the batches API essentially does transactions and would prefer it to use that API for transactions: https://developers.cloudflare.com/d1/worker-api/d1-database/#batch otherwise we have to manually write it

2. Durable Objects is becoming more popular and has its own database object alongside migrations pathways. This would require generating a SQL file that will be applied as migrations in text via constructor, and then using prisma driver to support the custom database object for Durable Objects. Honestly having raw SQL vs Prisma in the project (mix and match) feels prisma isn't ready for the cloudflare stack and the benefits don't outway the cost of the size for an ORM.

Assuming most folks will use this with cloudflare pages SSE which makes sense, but the above is my perspective using batches API and Durable Objects. Otherwise the extension works great.

**Comment by kevinsimper** [comment]
I am coming from Postgres and Prisma, and thought Cloudflare D1 + Prisma would be the same great experience.

But learning now that Cloudflare D1 has no way of disabling ON DELETE CASCADE, so if you are using default prisma, all your foreign keys will be deleted if you remove a column. 

That is not so ideal since with SQLite there is no way to DROP COLUMN or anything, so to remove a table you have to recreate. 

I think the default with Cloudflare D1 should be ON DELETE NoAction https://www.prisma.io/docs/orm/prisma-schema/data-model/relations/referential-actions#noaction

**Comment by rbasto1** [comment]
I've noticed that all my
```
"createdAt" DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP
```
fields have a fixed epoch value saved on them: `1970-01-01T00:00:00+00:00`
If I issue an INSERT INTO directly to the database omitting these fields, they get the correct value. So I'm thinking that the driver is doing something here to break this feature.
Is that a known bug or am I missing something in my setup?
